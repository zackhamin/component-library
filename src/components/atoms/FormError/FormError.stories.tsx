import { Story } from '@storybook/react';
import { FormError, FormErrorProps } from './FormError';

export default {
	title: 'atoms/FormError',
	component: FormError,
};

const Template: Story<FormErrorProps> = (args) => <FormError {...args} />;
const defaultProps = {
	children: 'Text',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
	isInvalid: true,
};
