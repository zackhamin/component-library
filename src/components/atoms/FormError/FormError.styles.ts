import styled from 'styled-components';
import { colors, typography } from 'Styles';

const StyledFormErrorMessage = styled.span`
	color: ${colors.error};
	font-size: ${typography.size.s2}rem;
	margin-top: 0.8rem;
	display: inline-block;
`;

export { StyledFormErrorMessage };
