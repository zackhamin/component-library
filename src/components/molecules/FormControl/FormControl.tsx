import { ConditionalFormProps } from '@Types/form';
import { createContext } from 'react';

interface FormControlContext {
	name?: string;
	id: string;
	isRequired?: boolean;
	isInvalid?: boolean;
	isSuccess?: boolean;
}

interface BaseFormControlProps extends FormControlContext {
	children: React.ReactNode;
	className?: string;
}

export type FormControlProps = BaseFormControlProps & ConditionalFormProps;

export const FormControlContext = createContext<FormControlContext | undefined>(
	undefined
);

const FormControl = ({
	children,
	name,
	id,
	isInvalid,
	isRequired,
	isSuccess,
	className,
}: FormControlProps) => {
	return (
		<FormControlContext.Provider
			value={{ name, id, isRequired, isInvalid, isSuccess }}
		>
			<div className={className}>{children}</div>
		</FormControlContext.Provider>
	);
};

export { FormControl };
