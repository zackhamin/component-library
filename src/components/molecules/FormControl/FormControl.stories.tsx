import { FormError } from 'Atoms/FormError';
import { FormLabel } from 'Atoms/FormLabel';
import { Story } from '@storybook/react';
import { TextInput } from 'Atoms/TextInput';
import { FormControl, FormControlProps } from './FormControl';

export default {
	title: 'molecules/FormControl',
	component: FormControl,
	argTypes: {
		isSuccess: {
			control: {
				type: 'boolean',
			},
		},
		isInvalid: {
			control: {
				type: 'boolean',
			},
		},
	},
};

const Template: Story<FormControlProps> = (args) => (
	<FormControl {...args}>
		<FormLabel>Label</FormLabel>
		<TextInput />
		<FormError>Error Message</FormError>
	</FormControl>
);
const defaultProps = {
	name: 'testName',
	id: 'testId',
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Invalid = Template.bind({});
Invalid.args = {
	...defaultProps,
	isInvalid: true,
};

export const Required = Template.bind({});
Required.args = {
	...defaultProps,
	isRequired: true,
};

export const Success = Template.bind({});
Success.args = {
	...defaultProps,
	isSuccess: true,
};
