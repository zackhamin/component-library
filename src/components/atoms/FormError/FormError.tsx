import { FormControlContext } from 'Molecules/FormControl';
import { StyledFormErrorMessage } from './FormError.styles';
import { useContext } from 'react';

export interface FormErrorProps {
	children: React.ReactNode;
	id?: string;
	isInvalid?: boolean;
}

const FormError = ({ children, id, isInvalid }: FormErrorProps) => {
	const ctx = useContext(FormControlContext);
	const formCtx = {
		id: ctx?.id || id,
		isInvalid: ctx?.isInvalid || isInvalid,
	};

	return (
		<div
			aria-live="polite"
			id={`${formCtx.id}--feedback`}
			data-testid="input-error"
		>
			{formCtx.isInvalid && (
				<StyledFormErrorMessage>{children}</StyledFormErrorMessage>
			)}
		</div>
	);
};

export { FormError };
