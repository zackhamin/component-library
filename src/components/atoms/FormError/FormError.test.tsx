import { FormControl } from 'Molecules/FormControl';
import { FormError } from '../FormError';

import { render, screen } from '@testing-library/react';

describe('FormError', () => {
	it('renders correctly', () => {
		const { container } = render(
			<FormError id="error">Error message</FormError>
		);
		expect(container).toMatchSnapshot();
	});

	it('renders empty live region by default', () => {
		render(<FormError id="error">Error message</FormError>);

		const alert = screen.getByTestId('input-error');

		expect(alert).toBeInTheDocument();
		expect(alert).toHaveAttribute('aria-live', 'polite');
		expect(alert).toHaveAttribute('id', 'error--feedback');
		expect(screen.queryByText('Error message')).not.toBeInTheDocument();
	});

	it('renders error message when isInvalid is directly set', () => {
		render(
			<FormError id="error" isInvalid>
				Error message
			</FormError>
		);

		expect(screen.getByTestId('input-error')).toBeInTheDocument();
		expect(screen.getByText('Error message')).toBeInTheDocument();
	});

	it('renders error message when invalid is set via form control', () => {
		render(
			<FormControl id="error" isInvalid>
				<FormError>Error message</FormError>
			</FormControl>
		);

		expect(screen.getByTestId('input-error')).toBeInTheDocument();
		expect(screen.getByText('Error message')).toBeInTheDocument();
	});

	it('correctly passes id to error message when set in form control', () => {
		render(
			<FormControl id="error" isInvalid>
				<FormError>Error message</FormError>
			</FormControl>
		);

		expect(screen.getByTestId('input-error')).toHaveAttribute(
			'id',
			'error--feedback'
		);
	});
});
