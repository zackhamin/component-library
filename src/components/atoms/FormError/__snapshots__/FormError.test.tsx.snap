// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormError renders correctly 1`] = `
<div>
  <div
    aria-live="polite"
    data-testid="input-error"
    id="error--feedback"
  />
</div>
`;
